staging/src/k8s.io/client-go/openapi/openapitest/fake.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapitest
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/client-go/openapi"
+)
+
+// NewFileOpenAPIClient returns a pointer to a testing
+// FileOpenAPIClient, which parses the OpenAPI GroupVersion
+// files in the passed directory path. Example GroupVersion
+// OpenAPI filename for apps/v1 would look like:
+//
+//	apis__apps__v1_openapi.json
+//
+// The current directories housing these hard-coded GroupVersion
+// OpenAPI V3 files is at:
+//
+//	<K8S_ROOT>/api/openapi-spec/v3
+//
+// An example to invoke this function for the test files:
+//
+//	NewFileOpenAPIClient("api/openapi-spec/v3")
+//
+// This function will search all subdirectories for .json files.
+func NewFileOpenAPIClient(dir string) (*FileOpenAPIClient, error) {
+	fullDirPath := filepath.Join("../../../../../../../", dir)
+	_, err := os.Stat(fullDirPath)
+	if os.IsNotExist(err) {
+		errStr := fmt.Sprintf("Directory does not exist: %s\n", fullDirPath)
+		return nil, errors.New(errStr)
+	}
+	values := map[string]openapi.GroupVersion{}
+	err = filepath.Walk(fullDirPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		filename := info.Name()
+		if !info.IsDir() && (filepath.Ext(filename) == ".json") {
```

Comment:
silently skips non-json but still valid open api documents like yaml or pb. Maybe should complain if a non supported type is found

Reply:
If you look at the current OpenAPI V3 files generated, there appear to be several unrelated files:
```
.well-known__openid-configuration_openapi.json
openid__v1__jwks_openapi.json
logs_openapi.json
```

There are several files which don't have versions (so wouldn't be useful for `GroupVersion.Schema()`):
```
apis__networking.k8s.io_openapi.json
```
versus
```
apis__networking.k8s.io__v1alpha1_openapi.json
apis__networking.k8s.io__v1_openapi.json
```

Not sure if we should try to print a warning, but I don't think it's worth throwing a panic for these situations.

Reply:
While I don't know if it makes sense for the `Root` interface in the other client, these files are actually provided by the apiserver (they are not "unrelated") and served at the path that we're looking at. so I think that's only fair to reproduce things as closely as they are from the actual API